### ISSUE TITLE
nuget pack -symbols: Missing source file error when the package contains XAML

### ISSUE THREAD

**Issue by tom-englert** [NuGet/Home#304]
A symbol package crated with "nuget pack -symbols" can't be uploaded to symbolsource.org if the package contains WPF/XAML controls. When analyzing or uploading the package it complains about missing files. 
When looking at the content of the package, only the MapTile.xaml.cs file is in the package, but the analyzer insists that MapTile.xaml and MapTile.g.cs should be included, too.
The source of the project where this happens can be downloaded at https://tomstoolbox.codeplex.com/SourceControl/latest

**Comment by yishaigalatzer** [comment]
Yes! We have the same problem with out symbol packages

**Comment by tom-englert** [comment]
Files to be added should be determined by looking at the PDB - not by trial and error!
(This would e.g. also fix other issues, like https://github.com/NuGet/Home/issues/529)

If NuGet claims to support SymbolSorce, it should do it right, or otherwise drop the entire feature.

There is a plugin for the PackageManger that shows these errors already at the client, (https://github.com/SymbolSource/SymbolSource.Community/tree/master/Sources/SymbolSource.Integration.NuGet.PackageExplorer), so you could easily adapt the code from there to provide a symbol package that is bullet proof.

**Comment by yishaigalatzer** [comment]
We would take a pr if you are interested. Currently this is low on our priority list.

**Comment by tom-englert** [comment]
https://nuget.codeplex.com/SourceControl/network/forks/tomenglert/private/contribution/8243

**Comment by yishaigalatzer** [comment]
Thank you!

I'm not sure if we can take dependency on this 3rd party package or not. We need to check legally if it is ok or not. Is the official package not sufficient?

Also please refrain from todo in the code and add tests.

The good news (I assume) is that you are unblocked, right?

**Comment by tom-englert** [comment]
I just wanted to use the same package as SymbolSource does (they are open source too, so the license should be fine) so I can reuse some of their code and be sure the result is identical. I will try the official CCI package and replace it if it works.

Todos; I'll remove the one about the code smell and create a bug instead, since I'm not sure why the code is there and if it is important.
I can't handle localization by my own, how is the correct workflow for this one?

I will check if I can add some useful tests.

**Comment by tom-englert** [comment]
The official CCI package is not usable - all the stuff needed is not public. Thats why the SymbolSource project created this shadow.
However I have cleaned up the references to minimize the dependencies to one assembly.

**Comment by yishaigalatzer** [comment]
I see, I will check and see if we can use it or not. Mind you the fact that an open source project is using a package, doesn't mean nuget.exe can.

Here is an alternative that works with Visual Studio (and might work with msbuild tools, need to verify)

https://github.com/nuproj/nuproj/blob/master/src/NuProj.Tasks/ReadPdbSourceFiles.cs

CC @tarrajobst

**Comment by terrajobst** [comment]
By the way, it's worth pointing out that there are cases where the PDB lists source files that aren't present at the time of building anymore. AFAIR I've hit this with VB compiler. So there still might be cases where SymbolSource.org might refuse a symbol package based on missing source files.

However, with NuProj I've found that using the PDB to determine which files to embed way more reliable than using the .csproj directly. There are way too many tools that generate temporary source files.

**Comment by tom-englert** [comment]
The COM component used in NuProj is registered in "C:\Program Files (x86)\Microsoft Visual Studio 12\Common7\Packages\Debugger\msdia120.dll" - so this is only available if you have Visual Studio installed. If you have only MsBuild tools installed, it won't work. Also every VisualStudio version provides a different one - msdia100 has another GUID, so if you want to support more than one VS version, the code has to be duplicated for every version:
![msdia](https://cloud.githubusercontent.com/assets/10229954/7985340/56fb488a-0ad0-11e5-8b21-023a96e20d1c.png)

The SymbolSource version of CCI does not state an explicit license, and since it's just a fork or the MS CCI, IMO it implicitly inherits the MS-PL license - so it should not make a difference. Since it's only a small part we are using, we could also just import the source files, if NuGet can go with MS-PL code.

**Comment by yishaigalatzer** [comment]
I don't get to make this kind calls. Will have to take this to the lawyers.

There is also the maintenance concern the cci package looks abandoned.

Perhaps we should make this a plugin.

I will need to get back to you

**Comment by yishaigalatzer** [comment]
Had a useful discussion with the Roslyn team on this, apparently there is a nice solution (which we will able to take post RTM.

Take a look at the code below:
https://github.com/dotnet/roslyn/tree/master/src/Tools/Source/Pdb2Xml
https://github.com/dotnet/roslyn/tree/master/src/Test/PdbUtilities/Pdb
http://source.roslyn.io/#Roslyn.Test.PdbUtilities/Pdb/PdbToXml.cs,4c28e20bd789be51

Here is a package with the interop definitions - http://www.nuget.org/packages/Microsoft.DiaSymReader/1.0.4-rc2 we can take it once it hits RTM.

Or we can find the interop definitions source.

CC @tmat

**Comment by terrajobst** [comment]
That's cool. We should probably change NuProj to use this one as well.

CC: @kovalikp, @aarnott

**Comment by Jaykul** [comment]
Is there any workaround for publishing projects containing XAML with nuget 2.8 or 3.x

**Comment by tom-englert** [comment]
@Jaykul: https://nuget.codeplex.com/SourceControl/network/forks/tomenglert/private/contribution/8243
That's what I am using and it works great.

**Comment by yishaigalatzer** [comment]
@tom-englert do you want to bring it to nuget2

**Comment by tom-englert** [comment]
@yishaigalatzer what repository should I use for nuget2? What about the license of https://www.nuget.org/packages/SymbolSource.Microsoft.Cci.Metadata/? Are we able to use this?

**Comment by yishaigalatzer** [comment]
1. nuget2
2. I provided alternatives to using that in the past, is there anything blocking you from using that suggestion?

**Comment by tom-englert** [comment]
1. OK, found that
2. I will have to investigate if one of the suggestions will be doing what we need.

**Comment by tom-englert** [comment]
The suggested DiaSymReader package has only interfaces, but no implementation.
I found  `Microsoft.DiaSymReader.Native 1.1.0-alpha2`, but have no idea how to use it.

So there is no usable alternative yet...

**Comment by tom-englert** [comment]
@yishaigalatzer I can add this to nuget2 as it is - but I won't spent any time on this without being sure that it can be integrated.

**Comment by yishaigalatzer** [comment]
I'm going to have someone show you the codbase that can be integrated, on this thread. And yes we would love to take it.

**Comment by tom-englert** [comment]
@yishaigalatzer any news about this?

**Comment by yishaigalatzer** [comment]
I thought we are waiting for you on this. What is blocking you?

Sent from my Windows Phone

**Comment by tom-englert** [comment]
I only have the version using SymbolSource.Microsoft.Cci.Metadata, but I don't have your go for using this, because of unresolved license questions.
There had been discussions about alternatives, but none of them is working.
On Sept. 4 your last comment was 

> I'm going to have someone show you the codbase that can be integrated, on this thread

So I'm still waiting on this.

**Comment by tmat** [comment]
@tom-englert I'd strongly recommend using Microsoft.DiaSymReader (interop interfaces) and Microsoft.DiaSymReader.Native (native x-copy deployable implementation of those interfaces, no need to do any COM registration) nuget packages -- the former already shipped, the latter will likely ship with Update 1. 

You can find examples of how to use the interfaces here:
https://github.com/dotnet/roslyn/tree/master/src/Test/PdbUtilities/Pdb

And feel free to ask questions.

**Comment by tom-englert** [comment]
So using an alpha pre-release (Microsoft.DiaSymReader.Native 1.1.0-alpha2) is OK?

**Comment by yishaigalatzer** [comment]
@tmat are we going to have an RTM version for this package any time soon? I much prefer relying on public RTM bits? Alternatively we can fork the code if its under the right license.

**Comment by tmat** [comment]
@yishaigalatzer When do you need it?

**Comment by yishaigalatzer** [comment]
If it was RTM within a month, that would work perfectly

**Comment by tmat** [comment]
hopefully it won't take that long.

**Comment by yishaigalatzer** [comment]
Then lets get on with this

**Comment by tom-englert** [comment]
@tmat https://github.com/dotnet/roslyn/tree/master/src/Test/PdbUtilities/Pdb does not use these interfaces, but https://github.com/dotnet/roslyn/tree/master/src/Debugging/Microsoft.DiaSymReader.PortablePdb, which I can't find anywhere as a package.

**Comment by tom-englert** [comment]
@yishaigalatzer @tmat any news on this, or is this a dead end?

**Comment by tmat** [comment]
@tom-englert It's certainly not dead end. We'll have an RC nupkg ready in a couple of days, with RTM planned for VS U1 RTM.

**Comment by tom-englert** [comment]
@tmat will this include the DiaSymReader.PortablePdb? Neither Microsoft.DiaSymReader nor Microsoft.DiaSymReader.Native provide anything of use to solve this!

**Comment by tmat** [comment]
@tom-englert Microsoft.DiaSymReader provides interfaces implemented by Microsoft.DiaSymReader.Native. You can use ISymUnmanagedReader.GetDocuments method to enumerate all source files referenced in the PDB.

**Comment by tom-englert** [comment]
@tmat so there are two questions left: 
- How do I create an object that implements ISymUnmanagedReader?
- How can I get Microsoft.DiaSymReader.Native - NuGet can't install it, it always complains there are no binaries for the selected framework?

**Comment by tmat** [comment]
project.json:

```
"dependencies": {
    "Microsoft.DiaSymReader.Native": "1.1.0-alpha2"
  }
```

.csproj:

```
  <ItemGroup>
    <Content Include="$(NuGetPackageRoot)\Microsoft.DiaSymReader.Native\1.1.0-alpha2\runtimes\win\native\Microsoft.DiaSymReader.Native.x86.dll">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <Visible>false</Visible>
    </Content>
    <Content Include="$(NuGetPackageRoot)\Microsoft.DiaSymReader.Native\1.1.0-alpha2\runtimes\win\native\Microsoft.DiaSymReader.Native.amd64.dll">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <Visible>false</Visible>
    </Content>
  </ItemGroup>
```

``` C#
public static class SymReaderFactory
    {
        [DefaultDllImportSearchPaths(DllImportSearchPath.AssemblyDirectory)]
        [DllImport("Microsoft.DiaSymReader.Native.x86.dll", EntryPoint = "CreateSymReader")]
        private extern static void CreateSymReader32(ref Guid id, [MarshalAs(UnmanagedType.IUnknown)]out object symReader);

        [DefaultDllImportSearchPaths(DllImportSearchPath.AssemblyDirectory)]
        [DllImport("Microsoft.DiaSymReader.Native.amd64.dll", EntryPoint = "CreateSymReader")]
        private extern static void CreateSymReader64(ref Guid id, [MarshalAs(UnmanagedType.IUnknown)]out object symReader);

        internal static ISymUnmanagedReader3 CreateNativeSymReader(Stream pdbStream, object metadataImporter)
        {
            object symReader = null;

            var guid = default(Guid);
            if (IntPtr.Size == 4)
            {
                CreateSymReader32(ref guid, out symReader);
            }
            else
            {
                CreateSymReader64(ref guid, out symReader);
            }

            var reader = (ISymUnmanagedReader3)symReader;
            int hr = reader.Initialize(metadataImporter, null, null, new ComStreamWrapper(pdbStream));
            SymUnmanagedReaderExtensions.ThrowExceptionForHR(hr);
            return reader;
        }
}
```

You don't need to supply a real metadata importer as long as you don't need to retrieve constant and variable signatures. A dummy implementation should suffice:

``` C#
 [ComImport, Guid("7DAC8207-D3AE-4c75-9B67-92801A497D44"), InterfaceType(ComInterfaceType.InterfaceIsIUnknown), TypeIdentifier]
public interface IMetadataImport   {  }

class DummyMetadataImport : IMetadataImport { }
```

**Comment by tom-englert** [comment]
@tmat OK, this seems to be the missing link - is there any official documentation for this?

**Comment by tom-englert** [comment]
@tmat now that uses an internal class 

roslyn/src/Compilers/Core/Portable/InternalUtilities/ComStreamWrapper.cs

Are you sure this was ever targeted for public usage?

Also with these interfaces i feel like falling back in the 1990s doing clumsy c++ COM interface programming. Can you provide a .Net wrapper that is just usable (like CCI was)?

**Comment by tmat** [comment]
@tom-englert Yes, the COM interfaces are documented (somewhat) on MSDN:

https://msdn.microsoft.com/en-us/library/ms233503(v=vs.110).aspx

Yes, it is targeted for public usage. The ComStreamWrapper is just an implementation of general IStream COM interface. You can copy it from Roslyn or write your own if you have some special requirements (not likely).

http://source.roslyn.io/#Microsoft.CodeAnalysis/InternalUtilities/ComStreamWrapper.cs,cb0b975fe0d90629

We could provide a wrapper. We just haven't had time to do so. 

We use some internal convenience helpers in Roslyn:
http://source.roslyn.io/#Roslyn.Test.PdbUtilities/Shared/SymUnmanagedReaderExtensions.cs,b6246ec4fee5b54a

Feel free to copy them as well.

**Comment by tmat** [comment]
@tom-englert Re clumsy C++ COM interfaces from 90's - I agree. We are dealing with a legacy proprietary format in the first place. We are working on a long term project to replace it with a new modern open format: https://github.com/dotnet/roslyn/blob/master/docs/specs/PortablePdb-Metadata.md, with modern reader and writer APIs.

**Comment by tom-englert** [comment]
@tmat, @yishaigalatzer another unresolved issue: If I modify the project as described above, $(NuGetPackageRoot) is undefined, and I had to replace it with a constant value. I found no documentation about how to set this dynamicaly.

**Comment by yishaigalatzer** [comment]
@tom-englert which project are you changing, where it doesn't work.
@tmat we should make the package work for net45 projects natively

**Comment by tom-englert** [comment]
@yishaigalatzer https://github.com/tom-englert/NuGet2/blob/2.9/src/CommandLine/CommandLine.csproj

**Comment by yishaigalatzer** [comment]
@tom-englert the PR proved a lot bigger than we thought, it doesn't actually work in the ilmerged nuget, and we have to restort to "trickery" to make it happen. @MeniZalzman has been spending a few days on this, and we didn't manage to get it working yet.

I'm unfortunately moving this to 3.4.

The good news is that the new symbol source now allows for packages to be pushed with missing files, so this is not as critical as it used to be.

**Comment by yishaigalatzer** [comment]
@MeniZalzman did you actually ever finish this?

**Comment by yishaigalatzer** [comment]
This was never finished, and we figured out the symbol source decided to become a lot more friendly. I'm closing as won't fix.

**Comment by tom-englert** [comment]
So what exactly is the problem with integrating this? It was not really much code, and I could build it and it worked exactly as expected.

**Comment by yishaigalatzer** [comment]
The main issues where
1. It wasn't from an official release - we cannot release a product from some unnoficial hack
2. It had no tests (as far as I recall)
3. It would break on none Windows machines.
4. It will not il merge into nuget.exe - That was where everything got a lot more complex.

**Comment by jainaashish** [comment]
This doesn't seem like a required feature now since symbol source allows packages to be published with missing files. So closing it for now.
